planner/core/rule_join_reorder_dp.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"math/bits"
+
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/sessionctx"
+)
+
+type joinReorderDPSolver struct {
+	ctx     sessionctx.Context
+	id2Node map[int]LogicalPlan
+	newJoin func(lChild, rChild LogicalPlan, eqConds []*expression.ScalarFunction) LogicalPlan
+}
+
+type joinGroupEdge struct {
+	nodeIDs []int
+	edge    *expression.ScalarFunction
+}
+
+func (s *joinReorderDPSolver) solve(joinGroup []LogicalPlan, conds []expression.Expression) (LogicalPlan, error) {
+	s.id2Node = make(map[int]LogicalPlan)
+	edges := make([][]int, len(joinGroup))
+	totalEdges := make([]joinGroupEdge, 0, len(conds))
+	addEdge := func(node1, node2 int, edgeContent *expression.ScalarFunction) {
+		totalEdges = append(totalEdges, joinGroupEdge{
+			nodeIDs: []int{node1, node2},
+			edge:    edgeContent,
+		})
+	}
+	// Build Graph for join group
+	for _, cond := range conds {
+		sf := cond.(*expression.ScalarFunction)
+		lCol := sf.GetArgs()[0].(*expression.Column)
+		rCol := sf.GetArgs()[1].(*expression.Column)
+		lIdx := findColumnIndexByGroup(joinGroup, lCol)
+		rIdx := findColumnIndexByGroup(joinGroup, rCol)
```

Comment:
So all the five lines are guaranteed by the caller?

Reply:
Now we assume that there're only join keys passing to this method.